File: src/camera/CameraClearanceQuery.js
import * as THREE from 'three';

export const CAMERA_CLEARANCE_DEFAULTS = Object.freeze({
  radius: 0.8,
  terrainClearance: 0.7,
  searchStep: 2.5,
  maxSearchRadius: 30,
  samplesPerRing: 16
});

function finiteVector(value) {
  return Number.isFinite(value?.x)
    && Number.isFinite(value?.y)
    && Number.isFinite(value?.z);
}

function readSize(size) {
  return {
    x: Number(size?.x ?? size?.width ?? 0),
    y: Number(size?.y ?? size?.height ?? 0),
    z: Number(size?.z ?? size?.depth ?? 0)
  };
}

function normalizeObstacle(obstacle) {
  if (!finiteVector(obstacle?.position)) return null;
  const size = readSize(obstacle.size);
  if (!(size.x > 0 && size.y > 0 && size.z > 0)) return null;
  return { position: obstacle.position, size, kind: obstacle.kind || 'obstacle' };
}

function sphereIntersectsBox(position, radius, obstacle) {
  const halfX = obstacle.size.x * 0.5;
  const halfY = obstacle.size.y * 0.5;
  const halfZ = obstacle.size.z * 0.5;
  const dx = Math.max(Math.abs(position.x - obstacle.position.x) - halfX, 0);
  const dy = Math.max(Math.abs(position.y - obstacle.position.y) - halfY, 0);
  const dz = Math.max(Math.abs(position.z - obstacle.position.z) - halfZ, 0);
  return dx * dx + dy * dy + dz * dz < radius * radius;
}

/**
 * Finds a deterministic collision-free camera origin. It uses lightweight
 * runtime obstacle snapshots rather than renderer raycasts, so the same query
 * can cover authored buildings, trees/scenery, vehicles, terrain, and water.
 */
export class CameraClearanceQuery {
  constructor({
    getTerrainHeight = () => 0,
    isWater = () => false,
    getObstacles = () => [],
    ...defaults
  } = {}) {
    if (typeof getTerrainHeight !== 'function') throw new TypeError('getTerrainHeight must be a function');
    if (typeof isWater !== 'function') throw new TypeError('isWater must be a function');
    if (typeof getObstacles !== 'function') throw new TypeError('getObstacles must be a function');
    this.getTerrainHeight = getTerrainHeight;
    this.isWater = isWater;
    this.getObstacles = getObstacles;
    this.defaults = Object.freeze({ ...CAMERA_CLEARANCE_DEFAULTS, ...defaults });
  }

  inspect(position, options = {}) {
    if (!finiteVector(position)) {
      return Object.freeze({ clear: false, reason: 'INVALID_POSITION', obstacle: null });
    }
    const radius = Math.max(0, Number(options.radius ?? this.defaults.radius) || 0);
    const terrainClearance = Math.max(
      radius,
      Number(options.terrainClearance ?? this.defaults.terrainClearance) || 0
    );
    const terrainHeight = Number(this.getTerrainHeight(position.x, position.z));
    const surfaceHeight = Number.isFinite(terrainHeight) ? terrainHeight : 0;
    if (position.y < surfaceHeight + terrainClearance) {
      return Object.freeze({
        clear: false,
        reason: 'TERRAIN',
        obstacle: null,
        minimumY: surfaceHeight + terrainClearance
      });
    }
    if (this.isWater(position)) {
      return Object.freeze({ clear: false, reason: 'WATER', obstacle: null });
    }

    const ignored = new Set(options.ignore || []);
    for (const candidate of this.getObstacles() || []) {
      if (ignored.has(candidate) || ignored.has(candidate?.entity)) continue;
      const obstacle = normalizeObstacle(candidate);
      if (obstacle && sphereIntersectsBox(position, radius, obstacle)) {
        return Object.freeze({ clear: false, reason: 'OBSTACLE', obstacle: candidate });
      }
    }
    return Object.freeze({ clear: true, reason: null, obstacle: null });
  }

  resolve(desiredPosition, options = {}) {
    if (!finiteVector(desiredPosition)) throw new TypeError('desiredPosition must contain finite x, y, and z');
    const resolved = desiredPosition.clone?.() || new THREE.Vector3(
      desiredPosition.x,
      desiredPosition.y,
      desiredPosition.z
    );
    this.#liftAboveTerrain(resolved, options);
    if (this.inspect(resolved, options).clear) return resolved;

    const step = Math.max(0.25, Number(options.searchStep ?? this.defaults.searchStep) || 0.25);
    const maxRadius = Math.max(step, Number(options.maxSearchRadius ?? this.defaults.maxSearchRadius) || step);
    const samples = Math.max(8, Math.round(Number(options.samplesPerRing ?? this.defaults.samplesPerRing) || 8));
    const preferredDirection = options.preferredDirection;
    const startAngle = finiteVector(preferredDirection)
      ? Math.atan2(preferredDirection.z, preferredDirection.x)
      : 0;

    for (let distance = step; distance <= maxRadius; distance += step) {
      for (let index = 0; index < samples; index += 1) {
        const angle = startAngle + (index / samples) * Math.PI * 2;
        resolved.set(
          desiredPosition.x + Math.cos(angle) * distance,
          desiredPosition.y,
          desiredPosition.z + Math.sin(angle) * distance
        );
        this.#liftAboveTerrain(resolved, options);
        if (this.inspect(resolved, options).clear) return resolved;
      }
    }

    throw new Error(`No safe camera position exists within ${maxRadius} world units.`);
  }

  #liftAboveTerrain(position, options) {
    const radius = Math.max(0, Number(options.radius ?? this.defaults.radius) || 0);
    const clearance = Math.max(
      radius,
      Number(options.terrainClearance ?? this.defaults.terrainClearance) || 0
    );
    const terrainHeight = Number(this.getTerrainHeight(position.x, position.z));
    position.y = Math.max(position.y, (Number.isFinite(terrainHeight) ? terrainHeight : 0) + clearance);
  }
}

export default CameraClearanceQuery;

File: godot/MetroPulse.Domain/Camera/CameraClearanceQuery.cs
namespace MetroPulse.Domain.Camera;

public static class CameraClearanceReasons
{
    public const string InvalidPosition = "INVALID_POSITION";
    public const string Terrain = "TERRAIN";
    public const string Water = "WATER";
    public const string Obstacle = "OBSTACLE";
}

public sealed record CameraClearanceDefaults(
    double Radius = 0.8,
    double TerrainClearance = 0.7,
    double SearchStep = 2.5,
    double MaximumSearchRadius = 30,
    int SamplesPerRing = 16);

public sealed record CameraObstacle(
    string Id,
    CameraVector3 Position,
    CameraVector3 Size,
    string Kind = "obstacle",
    string? EntityId = null);

public sealed record CameraClearanceOptions
{
    public double? Radius { get; init; }

    public double? TerrainClearance { get; init; }

    public double? SearchStep { get; init; }

    public double? MaximumSearchRadius { get; init; }

    public int? SamplesPerRing { get; init; }

    public CameraVector3? PreferredDirection { get; init; }

    public IReadOnlySet<string> IgnoredIds { get; init; } = new HashSet<string>(StringComparer.Ordinal);
}

public sealed record CameraClearanceInspection(
    bool Clear,
    string? Reason,
    CameraObstacle? Obstacle = null,
    double? MinimumY = null);

/// <summary>Deterministic collision-free camera-origin query over immutable world snapshots.</summary>
public sealed class CameraClearanceQuery
{
    private readonly Func<double, double, double> getTerrainHeight;
    private readonly Func<CameraVector3, bool> isWater;
    private readonly Func<IReadOnlyList<CameraObstacle>> getObstacles;
    private readonly CameraClearanceDefaults defaults;

    public CameraClearanceQuery(
        Func<double, double, double>? getTerrainHeight = null,
        Func<CameraVector3, bool>? isWater = null,
        Func<IReadOnlyList<CameraObstacle>>? getObstacles = null,
        CameraClearanceDefaults? defaults = null)
    {
        this.getTerrainHeight = getTerrainHeight ?? ((_, _) => 0);
        this.isWater = isWater ?? (_ => false);
        this.getObstacles = getObstacles ?? (() => Array.Empty<CameraObstacle>());
        this.defaults = defaults ?? new CameraClearanceDefaults();
    }

    public CameraClearanceInspection Inspect(
        CameraVector3 position,
        CameraClearanceOptions? options = null)
    {
        if (position is null || !position.IsFinite)
        {
            return new CameraClearanceInspection(false, CameraClearanceReasons.InvalidPosition);
        }
        options ??= new CameraClearanceOptions();
        double radius = NonNegative(options.Radius, defaults.Radius);
        double terrainClearance = Math.Max(radius, NonNegative(options.TerrainClearance, defaults.TerrainClearance));
        double terrainHeight = getTerrainHeight(position.X, position.Z);
        double surfaceHeight = double.IsFinite(terrainHeight) ? terrainHeight : 0;
        if (position.Y < surfaceHeight + terrainClearance)
        {
            return new CameraClearanceInspection(
                false,
                CameraClearanceReasons.Terrain,
                MinimumY: surfaceHeight + terrainClearance);
        }
        if (isWater(position)) return new CameraClearanceInspection(false, CameraClearanceReasons.Water);

        IReadOnlySet<string> ignored = options.IgnoredIds ?? new HashSet<string>(StringComparer.Ordinal);
        foreach (CameraObstacle candidate in getObstacles() ?? Array.Empty<CameraObstacle>())
        {
            if (candidate is null || ignored.Contains(candidate.Id)
                || (candidate.EntityId is not null && ignored.Contains(candidate.EntityId)))
            {
                continue;
            }
            if (NormalizeObstacle(candidate) is CameraObstacle obstacle
                && SphereIntersectsBox(position, radius, obstacle))
            {
                return new CameraClearanceInspection(false, CameraClearanceReasons.Obstacle, candidate);
            }
        }
        return new CameraClearanceInspection(true, null);
    }

    public CameraVector3 Resolve(
        CameraVector3 desiredPosition,
        CameraClearanceOptions? options = null)
    {
        ArgumentNullException.ThrowIfNull(desiredPosition);
        if (!desiredPosition.IsFinite)
        {
            throw new ArgumentException("desiredPosition must contain finite x, y, and z", nameof(desiredPosition));
        }
        options ??= new CameraClearanceOptions();
        CameraVector3 resolved = LiftAboveTerrain(desiredPosition, options);
        if (Inspect(resolved, options).Clear) return resolved;

        double step = Math.Max(0.25, Positive(options.SearchStep, defaults.SearchStep, 0.25));
        double maximumRadius = Math.Max(step, Positive(options.MaximumSearchRadius, defaults.MaximumSearchRadius, step));
        int samples = Math.Max(8, options.SamplesPerRing ?? defaults.SamplesPerRing);
        CameraVector3? preferred = options.PreferredDirection;
        double startAngle = preferred is not null && preferred.IsFinite
            ? Math.Atan2(preferred.Z, preferred.X)
            : 0;

        for (double distance = step; distance <= maximumRadius; distance += step)
        {
            for (int index = 0; index < samples; index += 1)
            {
                double angle = startAngle + ((double)index / samples) * Math.PI * 2;
                resolved = new CameraVector3(
                    desiredPosition.X + Math.Cos(angle) * distance,
                    desiredPosition.Y,
                    desiredPosition.Z + Math.Sin(angle) * distance);
                resolved = LiftAboveTerrain(resolved, options);
                if (Inspect(resolved, options).Clear) return resolved;
            }
        }
        throw new InvalidOperationException(
            $"No safe camera position exists within {maximumRadius} world units.");
    }

    private CameraVector3 LiftAboveTerrain(CameraVector3 position, CameraClearanceOptions options)
    {
        double radius = NonNegative(options.Radius, defaults.Radius);
        double clearance = Math.Max(radius, NonNegative(options.TerrainClearance, defaults.TerrainClearance));
        double terrainHeight = getTerrainHeight(position.X, position.Z);
        double surface = double.IsFinite(terrainHeight) ? terrainHeight : 0;
        return position with { Y = Math.Max(position.Y, surface + clearance) };
    }

    private static CameraObstacle? NormalizeObstacle(CameraObstacle obstacle)
    {
        if (!obstacle.Position.IsFinite || !obstacle.Size.IsFinite
            || obstacle.Size.X <= 0 || obstacle.Size.Y <= 0 || obstacle.Size.Z <= 0)
        {
            return null;
        }
        return obstacle;
    }

    private static bool SphereIntersectsBox(CameraVector3 position, double radius, CameraObstacle obstacle)
    {
        double dx = Math.Max(Math.Abs(position.X - obstacle.Position.X) - obstacle.Size.X * 0.5, 0);
        double dy = Math.Max(Math.Abs(position.Y - obstacle.Position.Y) - obstacle.Size.Y * 0.5, 0);
        double dz = Math.Max(Math.Abs(position.Z - obstacle.Position.Z) - obstacle.Size.Z * 0.5, 0);
        return dx * dx + dy * dy + dz * dz < radius * radius;
    }

    private static double NonNegative(double? value, double fallback)
    {
        double selected = value is not null && double.IsFinite(value.Value) ? value.Value : fallback;
        return Math.Max(0, selected);
    }

    private static double Positive(double? value, double fallback, double minimum)
    {
        double selected = value is not null && double.IsFinite(value.Value) && value.Value != 0
            ? value.Value
            : fallback;
        return Math.Max(minimum, selected);
    }
}
